colexec: incorrect result in ordered aggregator when batch size is 1

```
        --- FAIL: TestLogic/fakedist-vec-disk/vectorize (2.03s)
            logic.go:1158: overriding vectorize mode to on, setting batch size to 1
            logic.go:1835:

                testdata/logic_test/vectorize:1088: SELECT b, d, e, f, g, sum(a) FROM t45481 GROUP BY b, d, e, f, g ORDER BY b, d, e, f, g
                expected:
                    1  1  1  1  11111111-1111-1111-1111-111111111111  196
                    1  2  2  2  22222222-2222-2222-2222-222222222222  196
                    1  3  3  3  33333333-3333-3333-3333-333333333333  196
                    1  4  4  4  44444444-4444-4444-4444-444444444444  196
                    1  5  5  5  55555555-5555-5555-5555-555555555555  196
                    1  6  6  6  66666666-6666-6666-6666-666666666666  196
                    1  7  7  7  77777777-7777-7777-7777-777777777777  196
                    2  1  1  1  11111111-1111-1111-1111-111111111111  196
                    2  2  2  2  22222222-2222-2222-2222-222222222222  196
                    2  3  3  3  33333333-3333-3333-3333-333333333333  196
                    2  4  4  4  44444444-4444-4444-4444-444444444444  196
                    2  5  5  5  55555555-5555-5555-5555-555555555555  196
                    2  6  6  6  66666666-6666-6666-6666-666666666666  196
                    2  7  7  7  77777777-7777-7777-7777-777777777777  196
                    3  1  1  1  11111111-1111-1111-1111-111111111111  196
                    3  2  2  2  22222222-2222-2222-2222-222222222222  196
                    3  3  3  3  33333333-3333-3333-3333-333333333333  196
                    3  4  4  4  44444444-4444-4444-4444-444444444444  196
                    3  5  5  5  55555555-5555-5555-5555-555555555555  196
                    3  6  6  6  66666666-6666-6666-6666-666666666666  196
                    3  7  7  7  77777777-7777-7777-7777-777777777777  196
                    4  1  1  1  11111111-1111-1111-1111-111111111111  196
                    4  2  2  2  22222222-2222-2222-2222-222222222222  196
                    4  3  3  3  33333333-3333-3333-3333-333333333333  196
                    4  4  4  4  44444444-4444-4444-4444-444444444444  196
                    4  5  5  5  55555555-5555-5555-5555-555555555555  196
                    4  6  6  6  66666666-6666-6666-6666-666666666666  196
                    4  7  7  7  77777777-7777-7777-7777-777777777777  196
                    5  1  1  1  11111111-1111-1111-1111-111111111111  196
                    5  2  2  2  22222222-2222-2222-2222-222222222222  196
                    5  3  3  3  33333333-3333-3333-3333-333333333333  196
                    5  4  4  4  44444444-4444-4444-4444-444444444444  196
                    5  5  5  5  55555555-5555-5555-5555-555555555555  196
                    5  6  6  6  66666666-6666-6666-6666-666666666666  196
                    5  7  7  7  77777777-7777-7777-7777-777777777777  196
                    6  1  1  1  11111111-1111-1111-1111-111111111111  196
                    6  2  2  2  22222222-2222-2222-2222-222222222222  196
                    6  3  3  3  33333333-3333-3333-3333-333333333333  196
                    6  4  4  4  44444444-4444-4444-4444-444444444444  196
                    6  5  5  5  55555555-5555-5555-5555-555555555555  196
                    6  6  6  6  66666666-6666-6666-6666-666666666666  196
                    6  7  7  7  77777777-7777-7777-7777-777777777777  196
                    7  1  1  1  11111111-1111-1111-1111-111111111111  196
                    7  2  2  2  22222222-2222-2222-2222-222222222222  196
                    7  3  3  3  33333333-3333-3333-3333-333333333333  196
                    7  4  4  4  44444444-4444-4444-4444-444444444444  196
                    7  5  5  5  55555555-5555-5555-5555-555555555555  196
                    7  6  6  6  66666666-6666-6666-6666-666666666666  196
                    7  7  7  7  77777777-7777-7777-7777-777777777777  196

                but found (query options: "") :
                    1  3  1  1  11111111-1111-1111-1111-111111111111  196
                    1  4  3  3  33333333-3333-3333-3333-333333333333  196
                    1  5  4  4  44444444-4444-4444-4444-444444444444  196
                    1  6  5  5  55555555-5555-5555-5555-555555555555  196
                    1  1  6  6  66666666-6666-6666-6666-666666666666  196
                    1  7  2  2  22222222-2222-2222-2222-222222222222  196
                    1  6  7  7  77777777-7777-7777-7777-777777777777  196
                    2  1  6  6  66666666-6666-6666-6666-666666666666  196
                    2  2  1  1  11111111-1111-1111-1111-111111111111  196
                    2  3  2  2  22222222-2222-2222-2222-222222222222  196
                    2  4  3  3  33333333-3333-3333-3333-333333333333  196
                    2  5  4  4  44444444-4444-4444-4444-444444444444  196
                    2  7  5  5  55555555-5555-5555-5555-555555555555  196
                    2  3  7  7  77777777-7777-7777-7777-777777777777  196
                    3  2  1  1  11111111-1111-1111-1111-111111111111  196
                    3  4  3  3  33333333-3333-3333-3333-333333333333  196
                    3  5  2  2  22222222-2222-2222-2222-222222222222  196
                    3  6  5  5  55555555-5555-5555-5555-555555555555  196
                    3  7  4  4  44444444-4444-4444-4444-444444444444  196
                    3  7  6  6  66666666-6666-6666-6666-666666666666  196
                    3  1  7  7  77777777-7777-7777-7777-777777777777  196
                    4  2  1  1  11111111-1111-1111-1111-111111111111  196
                    4  3  2  2  22222222-2222-2222-2222-222222222222  196
                    4  4  3  3  33333333-3333-3333-3333-333333333333  196
                    4  4  7  7  77777777-7777-7777-7777-777777777777  196
                    4  5  4  4  44444444-4444-4444-4444-444444444444  196
                    4  6  5  5  55555555-5555-5555-5555-555555555555  196
                    4  1  6  6  66666666-6666-6666-6666-666666666666  196
                    5  2  1  1  11111111-1111-1111-1111-111111111111  196
                    5  3  2  2  22222222-2222-2222-2222-222222222222  196
                    5  5  3  3  33333333-3333-3333-3333-333333333333  196
                    5  6  5  5  55555555-5555-5555-5555-555555555555  196
                    5  2  6  6  66666666-6666-6666-6666-666666666666  196
                    5  7  4  4  44444444-4444-4444-4444-444444444444  196
                    5  1  7  7  77777777-7777-7777-7777-777777777777  196
                    6  3  1  1  11111111-1111-1111-1111-111111111111  196
                    6  4  3  3  33333333-3333-3333-3333-333333333333  196
                    6  5  4  4  44444444-4444-4444-4444-444444444444  196
                    6  1  5  5  55555555-5555-5555-5555-555555555555  196
                    6  6  2  2  22222222-2222-2222-2222-222222222222  196
                    6  7  6  6  66666666-6666-6666-6666-666666666666  196
                    6  2  7  7  77777777-7777-7777-7777-777777777777  196
                    7  3  2  2  22222222-2222-2222-2222-222222222222  196
                    7  4  1  1  11111111-1111-1111-1111-111111111111  196
                    7  5  4  4  44444444-4444-4444-4444-444444444444  196
                    7  6  3  3  33333333-3333-3333-3333-333333333333  196
                    7  6  6  6  66666666-6666-6666-6666-666666666666  196
                    7  7  5  5  55555555-5555-5555-5555-555555555555  196
                    7  7  7  7  77777777-7777-7777-7777-777777777777  196


            logic.go:2401:
                testdata/logic_test/vectorize:1141: error while processing
            logic.go:2401: testdata/logic_test/vectorize:1141: too many errors encountered, skipping the rest of the input
```

--- a/pkg/sql/colexec/aggregator_test.go
+++ b/pkg/sql/colexec/aggregator_test.go
@@ -444,6 +444,31 @@ func TestAggregatorMultiFunc(t *testing.T) {
 			},
 			name: "BoolAndOrBatch",
 		},
+		{
+			aggFns: []execinfrapb.AggregatorSpec_Func{
+				execinfrapb.AggregatorSpec_ANY_NOT_NULL,
+				execinfrapb.AggregatorSpec_ANY_NOT_NULL,
+				execinfrapb.AggregatorSpec_ANY_NOT_NULL,
+				execinfrapb.AggregatorSpec_MIN,
+				execinfrapb.AggregatorSpec_SUM_INT,
+			},
+			input: tuples{
+				{2, 1.0, "1.0", 2.0},
+				{2, 1.0, "1.0", 4.0},
+				{2, 2.0, "2.0", 6.0},
+			},
+			expected: tuples{
+				{2, 1.0, "1.0", 2.0, 6.0},
+				{2, 2.0, "2.0", 6.0, 6.0},
+			},
+			batchSize: 1,
+			colTypes:  []coltypes.T{coltypes.Int64, coltypes.Decimal, coltypes.Bytes, coltypes.Decimal},
+			name:      "MultiGroupColsWithPointerTypes",
+			groupCols: []uint32{0, 1, 2},
+			aggCols: [][]uint32{
+				{0}, {1}, {2}, {3}, {3},
+			},
+		},
 	}
 
 	for _, agg := range aggTypes {
@@ -452,7 +477,7 @@ func TestAggregatorMultiFunc(t *testing.T) {
 				if err := tc.init(); err != nil {
 					t.Fatal(err)
 				}
-				runTests(t, []tuples{tc.input}, tc.expected, unorderedVerifier,
+				runTestsWithTyps(t, []tuples{tc.input}, [][]coltypes.T{tc.colTypes}, tc.expected, unorderedVerifier,
 					func(input []Operator) (Operator, error) {
 						return agg.new(testAllocator, input[0], tc.colTypes, tc.aggFns, tc.groupCols, tc.aggCols, false /* isScalar */)
 					})

--- a/pkg/sql/colexec/any_not_null_agg_tmpl.go
+++ b/pkg/sql/colexec/any_not_null_agg_tmpl.go
@@ -200,14 +200,8 @@ func _FIND_ANY_NOT_NULL(a *anyNotNull_TYPEAgg, nulls *coldata.Nulls, i int, _HAS
 		// If we haven't seen any non-nulls for the current group yet, and the
 		// current value is non-null, then we can pick the current value to be the
 		// output.
-		// {{ if eq .LTyp.String "Bytes" }}
-		// Bytes type is special - we actually need to copy the value that we're
-		// getting in an "unsafe" way because col might be reused (and the
-		// underlying memory overwritten) on the next batches.
-		a.curAgg = append(a.curAgg[:0], execgen.UNSAFEGET(col, i)...)
-		// {{ else }}
-		a.curAgg = execgen.UNSAFEGET(col, i)
-		// {{ end }}
+		val := execgen.UNSAFEGET(col, i)
+		execgen.COPYVAL(a.curAgg, val)
 		a.foundNonNullForCurrentGroup = true
 	}
 	// {{end}}

--- a/pkg/sql/colexec/min_max_agg_tmpl.go
+++ b/pkg/sql/colexec/min_max_agg_tmpl.go
@@ -231,14 +231,15 @@ func _ACCUMULATE_MINMAX(a *_AGG_TYPEAgg, nulls *coldata.Nulls, i int, _HAS_NULLS
 	// {{ end }}
 	if !isNull {
 		if !a.foundNonNullForCurrentGroup {
-			a.curAgg = execgen.UNSAFEGET(col, i)
+			val := execgen.UNSAFEGET(col, i)
+			execgen.COPYVAL(a.curAgg, val)
 			a.foundNonNullForCurrentGroup = true
 		} else {
 			var cmp bool
 			candidate := execgen.UNSAFEGET(col, i)
 			_ASSIGN_CMP(cmp, candidate, a.curAgg)
 			if cmp {
-				a.curAgg = candidate
+				execgen.COPYVAL(a.curAgg, candidate)
 			}
 		}
 	}

--- a/pkg/sql/colexec/utils_test.go
+++ b/pkg/sql/colexec/utils_test.go
@@ -520,9 +520,26 @@ func setColVal(vec coldata.Vec, idx int, val interface{}) {
 			bytesVal = []byte(val.(string))
 		}
 		vec.Bytes().Set(idx, bytesVal)
-		return
+	} else if vec.Type() == coltypes.Decimal {
+		// setColVal is used in multiple places, therefore val can be either a float
+		// or apd.Decimal.
+		if decimalVal, ok := val.(apd.Decimal); ok {
+			vec.Decimal()[idx].Set(&decimalVal)
+		} else {
+			floatVal := val.(float64)
+			decimalVal, _, err := apd.NewFromString(fmt.Sprintf("%f", floatVal))
+			if err != nil {
+				execerror.VectorizedInternalPanic(
+					fmt.Sprintf("unable to set decimal %f: %v", floatVal, err))
+			}
+			// .Set is used here instead of assignment to ensure the pointer address
+			// of the underlying storage for apd.Decimal remains the same. This can
+			// cause the code that does not properly use execgen package to fail.
+			vec.Decimal()[idx].Set(decimalVal)
+		}
+	} else {
+		reflect.ValueOf(vec.Col()).Index(idx).Set(reflect.ValueOf(val).Convert(reflect.TypeOf(vec.Col()).Elem()))
 	}
-	reflect.ValueOf(vec.Col()).Index(idx).Set(reflect.ValueOf(val).Convert(reflect.TypeOf(vec.Col()).Elem()))
 }
 
 // opTestInput is an Operator that columnarizes test input in the form of tuples
@@ -967,6 +984,16 @@ func tupleEquals(expected tuple, actual tuple) bool {
 						continue
 					} else if !math.IsNaN(f1) && !math.IsNaN(f2) && math.Abs(f1-f2) < 1e-6 {
 						continue
+					}
+				}
+			}
+			if d1, ok := actual[i].(apd.Decimal); ok {
+				if f2, ok := expected[i].(float64); ok {
+					d2, _, err := apd.NewFromString(fmt.Sprintf("%f", f2))
+					if err == nil && d1.Cmp(d2) == 0 {
+						continue
+					} else {
+						return false
 					}
 				}
 			}

